core/trino-main/src/main/java/io/trino/sql/planner/iterative/rule/ReplaceJoinOverConstantWithProject.java

```diff
@@ -0,0 +1,180 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.trino.sql.planner.iterative.rule;
+
+import io.trino.matching.Captures;
+import io.trino.matching.Pattern;
+import io.trino.sql.planner.PlanNodeIdAllocator;
+import io.trino.sql.planner.Symbol;
+import io.trino.sql.planner.iterative.Rule;
+import io.trino.sql.planner.plan.Assignments;
+import io.trino.sql.planner.plan.JoinNode;
+import io.trino.sql.planner.plan.PlanNode;
+import io.trino.sql.planner.plan.ProjectNode;
+import io.trino.sql.planner.plan.ValuesNode;
+import io.trino.sql.tree.Expression;
+import io.trino.sql.tree.Row;
+
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+import static com.google.common.collect.Iterables.getOnlyElement;
+import static io.trino.sql.planner.optimizations.QueryCardinalityUtil.isAtLeastScalar;
+import static io.trino.sql.planner.optimizations.QueryCardinalityUtil.isAtMost;
+import static io.trino.sql.planner.plan.Patterns.join;
+import static io.trino.sql.tree.BooleanLiteral.TRUE_LITERAL;
+
+/**
+ * This rule transforms plans with join where one of the sources is
+ * a single-row ValuesNode, and the join condition is `on true`.
+ * The JoinNode is replaced with the other join source and a projection
+ * which appends constant values from the ValuesNode.
+ * This rule is similar to ReplaceRedundantJoinWithSource.
+ * <p>
+ * Note 1: When transforming an outer join (LEFT, RIGHT or FULL), and an
```

Comment:
If we had an "enforce at least one row" operation that produces nulls if there are no rows in the input, we could do this.

Reply:
That would be similar to `EnforceSingleRowNode` / `EnforceSingleRowOperator`. The main difficulty when adding it is to make sure that the new node doesn't block any optimization chains (pruning, pushdown etc.).